src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -24605,6 +24605,1467 @@ void Init()
                 );
         }
 
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void NullableValueType_RecursivePattern()
+        {
+            var c = CreateCompilation(@"
+#nullable enable
+
+public struct Test
+{
+    public int M(int? i)
+    {
+        switch (i)
+        {
+            case { HasValue: true }:
+                return i.Value;
+            default:
+                return i.Value;
+        }
+    }
+}
+");
+
+            c.VerifyDiagnostics(
+                // (10,20): error CS0117: 'int' does not contain a definition for 'HasValue'
+                //             case { HasValue: true }:
+                Diagnostic(ErrorCode.ERR_NoSuchMember, "HasValue").WithArguments("int", "HasValue").WithLocation(10, 20),
+                // (13,24): warning CS8629: Nullable value type may be null.
+                //                 return i.Value;
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "i").WithLocation(13, 24)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_RecursivePattern()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public struct Test
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+
+    public string M()
+    {
+        switch (this)
+        {
+            case { IsOk: true }:
+                return Value;
+            default:
+                return Value; // 1
+        }
+    }
+}
+", MemberNotNullWhenAttributeDefinition });
+
+            c.VerifyDiagnostics(
+                // (18,24): warning CS8603: Possible null reference return.
+                //                 return Value; // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReturn, "Value").WithLocation(18, 24)
+                );
+        }
+
+        [Fact, WorkItem(49750, "https://github.com/dotnet/roslyn/issues/49750")]
+        public void MemberNotNullWhenTrue_RecursivePattern_Nested()
+        {
+            var c = CreateCompilation(new[] { @"
+using System.Diagnostics.CodeAnalysis;
+#nullable enable
+
+public class C
+{
+    [MemberNotNullWhen(true, nameof(Value))]
+    public bool IsOk { get; }
+    public string? Value { get; }
+}
+public class Container
+{
+    public C? Item;
+
+    public string M()
+    {
+        switch (this)
+        {
+            case { Item: { IsOk: true } item }:
+                return item.Value;
```

Comment:
should we have a version of this test which returns `this.Item.Value` here?

Reply:
Added test above (`MemberNotNullWhenTrue_RecursivePattern_Nested`)


---
In reply to: [574034771](https://github.com/dotnet/roslyn/pull/51001#discussion_r574034771) [](ancestors = 574034771)